static/coffee/flows/services.coffee

```diff
@@ -741,17 +741,49 @@ app.factory 'Flow', ['$rootScope', '$window', '$http', '$timeout', '$interval',
     isPausingRulesetType: (ruleset_type) ->
       return ruleset_type in ['wait_message', 'wait_recording', 'wait_digit', 'wait_digits']
 
+    isTargetAllowedForInterrupt: (targetId) ->
+      node = @getNode(targetId)
+
+      if node?.rules
+        return false
+
+      if node?.actions
+        for action in node.actions
+          if action.type in ['end_ussd']
+            return false
+      return true
+
+    detectInterruptRule: (sourceId, ruleId) ->
+      node = @getNode(sourceId)
+
+      if node?.actions
+        return
+
+      for rule in node.rules
+        if rule.uuid == ruleId and rule.test.type == 'interrupted_status'
+          return true
+
+    checkSourceForInterrupt: (actionset) ->
+      connections = Plumb.getConnectionMap({ target: actionset.uuid })
+
+      for source of connections
+        uuids = source.split('_')
+        sourceId = uuids[0]
+        ruleId = uuids[1]
+        if @detectInterruptRule(sourceId, ruleId)
+          @updateDestination(source, null)
+
     # check if a potential connection would result in an invalid loop
     detectLoop: (nodeId, targetId, path=[]) ->
 
       node = @getNode(targetId)
 
       # can't go back on ourselves
-      if nodeId == targetId and not @isUssdRuleset(node)
+      if nodeId == targetId and (not @isUssdRuleset(node) or @isUssdRuleset(node) and not window.ussd_push_enabled)
```

Comment:
These conditions are getting a bit crazy. Can you simplify and/or make this a method?